### PR TITLE
Split SliceAttr into positional and named arguments, to be reusable in other operators

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- Refactoring to improve reusability of SliceAttr logic

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Split SliceAttr implementation into two: NamedSliceAttr and PositionalSliceAttr, so that the named arguments logic can be used in other operators dealing with ROIs*
 - Affected modules and functionalities:
     *Slice and Slice decoders*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *Existing tests apply*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-2127]*

### CODE REVIEW

**Review by awolant — APPROVED** [review]

File: dali/operators/generic/slice/slice_attr.h [hunk 1]
```diff
@@ -273,24 +338,55 @@ class SliceAttr {
...
+            "when using positional slice arguments");
+      }
+      DALI_ENFORCE(ws.NumInput() <= 1,
+                  "Named arguments start/end/shape are not compatible with positional"
+                  " anchor and shape inputs");
+    } else {
+      if (ws.NumInput() != 3) {
+        DALI_FAIL(
+            make_string("Expected named slice arguments (e.g. start/end, start/shape) "
+                        "or positional inputs start, shape. Got ",
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:368]: Nitpick:
```suggestion
                        "or positional inputs (start, shape). Got ",
```

File: dali/operators/generic/slice/slice_attr.h [hunk 2]
```diff
@@ -273,24 +338,55 @@ class SliceAttr {
...
+      if (spec_.HasArgument("normalized_anchor") || spec_.HasArgument("normalized_shape")) {
+        DALI_WARN(
+            "Warning: ``normalized_anchor``/``normalized_shape`` is only relevant "
+            "when using positional slice arguments");
+      }
+      DALI_ENFORCE(ws.NumInput() <= 1,
+                  "Named arguments start/end/shape are not compatible with positional"
+                  " anchor and shape inputs");
+    } else {
+      if (ws.NumInput() != 3) {
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:365]: Perhaps the number of data inputs should be a member?
```suggestion
      DALI_ENFORCE(ws.NumInput() <= NumDataInputs(),
                  "Named arguments start/end/shape are not compatible with positional"
                  " anchor and shape inputs");
    } else {
      if (ws.NumInput() != NumDataInputs()) {
```
where
```
  int NumDataInputs() const {
    return spec_.GetSchema().MinNumInput();
  }
````

File: dali/operators/generic/slice/slice_attr.h [hunk 3]
```diff
@@ -199,6 +184,86 @@ class SliceAttr {
...
+    auto curr_batch_size = ws.GetInputBatchSize(0);
+    int ndim = ws.GetInputDim(0);
+
+    auto args_shape = TensorShape<1>(ndim);
+    if (!axes_.empty() || !axis_names_.empty()) {
+      args_shape[0] = std::max(static_cast<int>(axes_.size()),
+                               static_cast<int>(axis_names_.size()));
+    }
+
+    if (ws.NumInput() != 3)
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:239]: Use `spec_.GetSchema().MinNumInput() + 2`?

File: dali/operators/generic/slice/slice_attr.h [hunk 4]
```diff
@@ -273,24 +338,55 @@ class SliceAttr {
...
+  template <typename Backend>
+  void ProcessArguments(const workspace_t<Backend> &ws) {
+    named_args_ = named_slice_attr_.template ProcessArguments<Backend>(ws);
+    if (named_args_) {
+      if (spec_.HasArgument("normalized_anchor") || spec_.HasArgument("normalized_shape")) {
+        DALI_WARN(
+            "Warning: ``normalized_anchor``/``normalized_shape`` is only relevant "
+            "when using positional slice arguments");
+      }
+      DALI_ENFORCE(ws.NumInput() <= 1,
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:361]:
```suggestion
      DALI_ENFORCE(ws.NumInput() == spec_.GetSchema().MinNumInput(),
```

File: dali/operators/generic/slice/slice_attr.h (hunk 2 above)
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:365]: I think this check belongs to positional SliceAttr - and it even is there.

File: dali/operators/generic/slice/slice_attr.h [hunk 5]
```diff
@@ -273,24 +338,55 @@ class SliceAttr {
...
+    else
+      return pos_slice_attr_.GetCropWindowGenerator(data_idx);
+  }
 
-  bool has_start_, has_end_, has_shape_;
+ private:
+  const OpSpec &spec_;
+  NamedSliceAttr named_slice_attr_;
+  PositionalSliceAttr pos_slice_attr_;
+  bool named_args_ = false;
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:386]:
```suggestion
  bool has_named_args_ = false;
```

File: dali/operators/generic/slice/slice_attr.h [hunk 6]
```diff
@@ -273,24 +338,55 @@ class SliceAttr {
...
+  explicit inline SliceAttr(const OpSpec &spec)
+      : spec_(spec),
+        named_slice_attr_(spec),
+        pos_slice_attr_(spec) {
+  }
 
-  std::vector<CropWindowGenerator> crop_window_generators_;
+  template <typename Backend>
+  void ProcessArguments(const workspace_t<Backend> &ws) {
+    named_args_ = named_slice_attr_.template ProcessArguments<Backend>(ws);
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:354]:
```suggestion
    use_named_args_ = named_slice_attr_.template ProcessArguments<Backend>(ws);
```

File: dali/operators/generic/slice/slice_attr.h [hunk 7]
```diff
@@ -199,6 +184,86 @@ class SliceAttr {
...
+};
+
+
+class PositionalSliceAttr {
+ public:
+  explicit inline PositionalSliceAttr(const OpSpec &spec)
+      : spec_(spec),
+        normalized_anchor_(spec.GetArgument<bool>("normalized_anchor")),
+        normalized_shape_(spec.GetArgument<bool>("normalized_shape")),
+        crop_window_generators_(spec.GetArgument<int>("max_batch_size")) {
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:214]: Nitpick: I think that initialization of non-const fields of fundamental type is more readable when it's done by simple assignment in the constructor body.

File: dali/operators/generic/slice/slice_attr.h [hunk 8]
```diff
@@ -199,6 +184,86 @@ class SliceAttr {
...
+  bool has_start_, has_end_, has_shape_;
+};
+
+
+class PositionalSliceAttr {
+ public:
+  explicit inline PositionalSliceAttr(const OpSpec &spec)
+      : spec_(spec),
+        normalized_anchor_(spec.GetArgument<bool>("normalized_anchor")),
+        normalized_shape_(spec.GetArgument<bool>("normalized_shape")),
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:213]: Nitpick: I think that initialization of non-const fields of fundamental type is more readable when it's done by simple assignment in the constructor body.

File: dali/operators/generic/slice/slice_attr.h (hunk 7 above)
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:214]: Likewise, this is harder to read than just calling resize in the body and the performance benefits are small if any and in this case they are of no concern (this function is called once per operator life time).

File: dali/operators/generic/slice/slice_attr.h [hunk 9]
```diff
@@ -33,28 +33,36 @@
...
-  explicit inline SliceAttr(const OpSpec &spec)
+  explicit inline NamedSliceAttr(const OpSpec &spec,
+                                 const char* kStartName = "start",
+                                 const char* kRelStartName = "rel_start",
+                                 const char* kEndName = "end",
+                                 const char* kRelEndName = "rel_end",
+                                 const char* kShapeName = "shape",
+                                 const char* kRelShapeName = "rel_shape",
+                                 const char* kAxesName = "axes",
+                                 const char* kAxisNamesName = "axis_names")
```
Comment (mzient) [inline dali/operators/generic/slice/slice_attr.h:46]: These parameters are not constant expressions, as the name suggests and the logic of this function doesn't require them to outlive the call.
```suggestion
                                 const char* start_name = "start",
                                 const char* rel_start_name = "rel_start",
                                 const char* end_name = "end",
                                 const char* rel_end_name = "rel_end",
                                 const char* shape_name = "shape",
                                 const char* rel_shape_name = "rel_shape",
                                 const char* axes_name = "axes",
                                 const char* axis_names_name = "axis_names")
```

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**